tests/python/pants_test/rules/test_test.py

```diff
@@ -124,3 +127,66 @@ def test_coordinator_python_test(self):
       result,
       AddressAndTestResult(addr, TestResult(status=Status.FAILURE, stdout='foo', stderr=''))
     )
+
+  def test_globbed_test_target(self):
+    bfaddr = BuildFileAddress(None, 'tests', 'some/dir')
+    target_adaptor = PythonTestsAdaptor(type_alias='python_tests')
+    with self.captured_logging(logging.INFO):
+      result = run_rule(
+        coordinator_of_tests,
+        HydratedTarget(bfaddr.to_address(), target_adaptor, ()),
+        UnionMembership(union_rules={TestTarget: [PythonTestsAdaptor]}),
+        AddressProvenanceMap(bfaddr_to_spec={
+          bfaddr: DescendantAddresses(directory='some/dir')
+        }),
+        {
+          (TestResult, PythonTestsAdaptor):
+            lambda _: TestResult(status=Status.SUCCESS, stdout='foo', stderr=''),
+        })
+
+      self.assertEqual(
+        result,
+        AddressAndTestResult(bfaddr.to_address(),
+                             TestResult(status=Status.SUCCESS, stdout='foo', stderr=''))
+      )
+
+  def test_globbed_non_test_target(self):
+    bfaddr = BuildFileAddress(None, 'bin', 'some/dir')
+    target_adaptor = PythonBinaryAdaptor(type_alias='python_binary')
+    with self.captured_logging(logging.INFO):
+      result = run_rule(
+        coordinator_of_tests,
+        HydratedTarget(bfaddr.to_address(), target_adaptor, ()),
+        UnionMembership(union_rules={TestTarget: [PythonTestsAdaptor]}),
+        AddressProvenanceMap(bfaddr_to_spec={
+          bfaddr: DescendantAddresses(directory='some/dir')
+        }),
+        {
+          (TestResult, PythonTestsAdaptor):
+            lambda _: TestResult(status=Status.SUCCESS, stdout='foo', stderr=''),
+        })
+
+      self.assertEqual(
+        result,
+        AddressAndTestResult(bfaddr.to_address(), None)
+      )
+
+  def test_single_non_test_target(self):
+    bfaddr = BuildFileAddress(None, 'bin', 'some/dir')
+    target_adaptor = PythonBinaryAdaptor(type_alias='python_binary')
+    with self.captured_logging(logging.INFO):
+      # Note that this is not the same error message the end user will see, as we're resolving
```

Comment:
Yeah, I don't love this either. But I'm pretty against integration tests for very specific bits of functionality. They have incredibly high overhead and we should be using them much more sparingly than we do. In so many cases they exist because some underlying functionality is too hard to test in isolation. 

I'd prefer that we figure out a way to have run_rule apply the same union casting logic as in prod, and get the right error message that way.